iceoryx_utils/include/iceoryx_utils/cxx/forward_list.hpp

```diff
@@ -47,9 +48,9 @@ namespace cxx
 ///
 ///
 /// @param T type user data to be managed within list
-/// @param Capacity number of maximum list elements a client can push to the list. minimum value is '1'
+/// @param Capacity number of maximum list elements a client can push to the list.
 template <typename T, uint64_t Capacity>
-class forward_list
+class forward_list : public SizedUninitializedArray<T, Capacity>
```

Comment:
You are right, this should be implemented as composition instead of inheritance. So I would change to 
- `class UninitializedArray<T, Capacity>`
- `class SizedContainer<Capacity>`
- vector, list, and forward_list use private members of `UninitializedArray` and `SizedContainer`. Only applicable public methods are added (e.g., vector won't have set_size() anymore).
@elBoberido @elfenpiff Are you both fine with this proposal?

Reply:
@elBoberido @elfenpiff: Does my previous proposal solve the review findings? If yes, I would update the PR, accordingly.

Reply:
@sschnitzer sorry, this totally slipped my attention.

I still think that a specialization for `UninitializedArray` for capacity of 0 is not necessary, since the final struct still has a size of 1. But if you want to do the work I won't stop you from doing it.
If you really want to save some space, then `SizeValue` must contain the `UninitializedArray` and then calling it `SizedContainer`, `SizedStorage`, `FixedSizeDynamicContainer` or `WhatChamaCallitStorage` makes sense again. The specialization of this container makes sense in order to have a member with only one byte, which results in a `cxx::vector` with a size of also only one byte.
Just to prevent misunderstandings. I proceeded on the assumption your suggestion was not to have `SizeValue` as member of `UninitializedArray`.

Also, as elfenpiff already pointed out. For the implementation of STL types we use the STL naming convention. I think we need to clarify this in the CONTRIBUTION.md. This is not part of the public API, but just an implementation detail. Since @elfenpiff implemented most of the STL container, do we haven an unwritten rule to also have this naming convention for internal constructs?

Reply:
There are two options remaining:
a) Keep  `UninitializedArray` and `SizeValue` separate. 
    - Pro: allows to have an UninitializedArray without a size, or a size of different data type (https://github.com/eclipse-iceoryx/iceoryx/pull/491#discussion_r560125879)
    - Pro: The array-related methods are implemented just once
b) Merge the two into something like `SizedUninitializedArray`
    - Pro: Saves 1 byte for Capacity 0, since a `SizeValue` is not a separate empty struct consuming 1 Byte
    - Pro: one class less (.hpp + .inl)

In case of a), I also would prefer the "trick":
```
using element_t = alignas(Capacity == 0 ? 1 : alignof(T)) uint8_t[Capacity == 0 ? 1 : sizeof(T)];
element_t m_data[Capacity == 0 ? 1 : Capacity];
```
In case of b), the trick is not needed. For capacity 0, the `end()` iterator would just return `this()`.

If you ask me, I prefer option b), since it is slightly more memory-efficient and the loss in modularity might be a case of YAGNI.

Reply:
There is option c)
- have a `ContainerStorage` which inherits from `UninitializedArray` and has a `fillLevel` member if capacity is not 0
- this has all the methods from `UninitializedArray`
- the number of specialized methods is minimal <- this is the main reason I'm leaning towards this solution
- since this is not user facing, inheritance shouldn't create problems further down the road

Currently I can think only of the `cxx::vector` where this member is useful. If the `cxx::list` also uses it, it's a source for future optimization if we want to have the container as small as possible.
There are other container in our code base which could make use of the `UninitializedArray` but don't need the additional size member because they handle this stuff differently, e.g. the `SoFi`  or the `ObjectPool`. For these, the size member would be dead memory.

Reply:
Right, Option c) has the same Pro's as a) and additionally the first Pro from b). 
Btw., the `SoFi` and the `ObjectPool` also have `m_size`, so this could also be refactored to use ContainerStorage.
@elfenpiff Do you agree to proceed with Option c)?

Reply:
@sschnitzer the `m_size` from the `SoFi` has a different meaning. A better name would be `m_currentMaxCapacity`. This was introduced to have a somewhat resizable `SoFi`. It just limits the the number of elements the `SoFi` can store, but doesn't tell anything about the current fill level.
With the `ObjectPool` you are right.

I also just noticed that @MatthiasKillat did something similar for his queue
https://github.com/eclipse-iceoryx/iceoryx/blob/master/iceoryx_utils/include/iceoryx_utils/internal/concurrent/lockfree_queue/buffer.hpp
Maybe we can reuse that and put it into a more common place

Reply:
@elBoberido It should probably not be called buffer then, unitialized_array might be better, I have no other proposals. As far as I am not sure whether it can allocate 0 bytes out of the box. Using it with template parameter 0 will cause an internal array to have size 0, which is not portable. I think I used it (in a test) with capacity 0 though.

This could easily be adapted to allocate 1 byte internally if Capacity is 0, it is just compile time computation.
As for where it should be placed I do not know, currently it is just a helper structure for the lockfree queue.

Reply:
@elBoberido @elfenpiff I updated the PR according to Option c). Can you please have a look?